### DISCUSSION TITLE
Correct working flow of accept()

### DISCUSSION THREAD

**Discussion by JaesukLim** [ANLAB-KAIST/KENSv3#177] [Project Q&A 2. Transport Layer (TCP)] [answered]
I am confused about when the connection request on the queue pops.
Reading from past discussion, KENS' backlog is saving pre-handshake requests.
What I thought is, 
Client SYN packet -> pushed to the queue -> send SYNACK -> receive ACK -> accept() -> pop the request
But this is actually save the established connections. Then how about following order?
Clicne SYN packet -> pushed to the queue -> accept() (block) -> pop the request -> send SYNACK -> receive ACK -> return accept() with file descriptor

**Comment by Pusnow** [comment, marked as answer]
For KENS, I recommend following order (flow before Linux 2.2):

```
Client SYN packet -> pushed to the queue -> send SYNACK -> receive ACK -> pop the request ->  accept()
```


<details>
  <summary>Real implementations in modern OSes. You don't need to read this.</summary>

FYI, the above is not the exact flow of modern OSes.
In the real implementations, the flow goes like below:

```
Client SYN packet -> pushed to the queue -> send SYNACK -> receive ACK -> accept() -> pop the request 
```

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/Documentation/man/listen.2.txt#L60-L64

</details>

**Reply by JaesukLim** [reply]
Thank you for quick response!
Then the accept() is blocked only for waiting next packet when the pending queue is empty?
I still confusing with the blocking behavior of accept()

**Reply by Pusnow** [reply]
> Then the accept() is blocked only for waiting next packet when the pending queue is empty?

Yes, `accept()` is blocked until a new connection is established if the queue is empty.
